PM: Refactor S2RAM kconfigs to rely on devicetree and soc

bjarki-andreasen opened the pull request:
TL;DR: Make `CONFIG_PM_S2RAM` and `CONFIG_PM_S2RAM_HAS_CUSTOM_MARKING` automatically selected, so users don't need to manually select them.

The `CONFIG_PM_S2RAM` config is currently user selectable in Kconfig, even though the SoC may not have any enabled `suspend-to-ram` state in the devicetree. Similarly, the devicetree may have an enabled `suspend-to-ram` state, and `CONFIG_PM` enabled, but `CONFIG_PM_S2RAM` be disabled, in which case PM will try to enter the state, and consistently fail since the implementation is not available. This is confusing to users. If `CONFIG_PM` is enabled, one or more `suspend-to-ram` states are enabled in the devicetree, surely they should be available.

The first commit refactors the `CONFIG_PM_S2RAM` to be promtless, and depend on the presence of any enabled `suspend-to-ram` states in the devicetree. Users can now rely on any enabled sleep state in the devicetree to be available if `CONFIG_PM` is selected.

The `CONFIG_PM_S2RAM_HAS_CUSTOM_MARKING` config is also currently user selectable in Kconfig, even though this option is not optional for the SoCs which depend on it for their PM_S2RAM implementations, and the option will break any PM_S2RAM implementations which rely on the "default" marking.

The second commit replaces the `CONFIG_PM_S2RAM_HAS_CUSTOM_MARKING` option with the promtless `CONFIG_PM_HAS_S2RAM_CUSTOM_MARKING`, which is selected by the SoCs if they need it for their PM_S2RAM implementations. Users no longer need to care about this option.

After these two refactors, users only need to enable `CONFIG_PM` to enable power management, and can disable any sleep states they don't want to use in the devicetree.

bjarki-andreasen-nordic committed:
pm: bind the PM_S2RAM kconfig option to devicetree

Currently it is possible to have an enabled (status="okay")
"suspend-to-ram" "zephyr,power-state" in the devicetree,
CONFIG_PM=y, but CONFIG_PM_S2RAM=n

This means the presence and state of the "suspend-to-ram" power
state in the devicetree does not match the PM_S2RAM option,
despite PM_S2RAM being dependent on the devicetree state.

This commit makes the devicetree the "source of truth" for
whether one or more s2ram power states shall be included and
supported, by enabling PM_S2RAM if any s2ram power state with
status "okay" is present in the devicetree. To disable the
s2ram power state, like with any other state, disable it by
setting its state to "disabled".

The help of the now promptless PM_S2RAM config has been
updated to reflect the new meaning. It previously held
cortex-m specific implementation details, these have been
removed as these details are already present in the file
pointed to as well (arch/arm/core/cortex_m/pm_s2ram.c
line 22)

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
pm: refactor PM_S2RAM_CUSTOM_MARKING option to be promptless

The config PM_S2RAM_CUSTOM_MARKING is not an optional config for a
user to select, it is required by some soc implementations of S2RAM,
in which case it must be selected by the soc.

Refactor the configuration to be HAS_PM_S2RAM_CUSTOM_MARKING, and
make the currently only soc which needs it select it. Then update
samples which previously had to select this option for this soc.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

erwango reviewed (changes requested):
In general I think this goes too much in the direction of activating a SW feature through dt configuration, I'd like others to comment on that (@decsny for instance).
If it was to be accepted, some clarification would be required to make the possibility of control more explicit to the user.

bjarki-andreasen-nordic committed:
doc: releases: document PM_S2RAM and PM_S2RAM_CUSTOM_MARKING refactor

Document the refactoring of the PM_S2RAM and PM_S2RAM_CUSTOM_MARKING
in the release notes and migration guide for release 4.3

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>
